IMA-WorldHealth/bhima
# Added checking of unit prices on stock entry

One of the problems of dealing with the stock is dealing with inventory article units correctly during data entry.  It easy to make a mistake if grams are mistaken for kilograms, etc.   This PR adds checking during the data entry to make sure the the specified unit price makes sense.    Currently it will complain if the proposed unit cost is less than 50% of the median purchase unit price or more than twice the median purchase price.

This PR was also merged with a different branch that does very similar sanity checking while creating purchase orders.

Closes https://github.com/IMA-WorldHealth/bhima/issues/6052

This PR also does some minor improvements to the purchase price analysis report.  It adds the median unit cost in the summary area at the bottom of the report.  It also fixes an error in the constants for the underlying SQL query.  (3 was used for flux FROM_INTEGRATION instead of the correct 13; Now this SQL query uses the server-side constants file for those values).

**TESTING**
- Suggest using the Vanga dataset
- Select inventory article with a history of purchases
   - Suggest:  107-094 Eau Distillée
   - Try its purchase price analysis (in the action menu).  Note the median unit cost at the summary at the bottom.
- Create a purchase order with this inventory article.  Use a unit cost similar to the median cost but not the same value
- In the purchase registry, change the status of the purchase order to "Confirmed"
- Go to the Stock Entry page
- Select a depot that supports stock entry from Purchase
- Select from Purchase and select the purchase order you just created
- Click on the "Lots" indicator and chose a lot
- On the unit cost line try entering a very small number.  When you exit the entry cell, you should receive a warning dialog pop-up.   If you Confirm, the new unit cost will be accepted.  If you Cancel, the previous unit costs will be restored and a warning will be added at the bottom of the modal.
- Try the same thing with a unit price more than double the median cost.

- Try a similar sequence where you try creating a purchase order with too large or too small a value.  
- Also try creating a purchase order for an inventory article with no prior purchase history, confirm the purchase, and try stock entry for it (high and low).
